File: shelf/src/test/java/com/toddway/shelf/CacheThenNewTests.java
package com.toddway.shelf;

import org.junit.Before;
import org.junit.Test;

import java.io.File;
import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.observers.TestObserver;

import static org.junit.Assert.assertEquals;

/**
 * Created by tway on 1/9/17.
 */

public class CacheThenNewTests {

    ShelfItem item;
    TestObserver<String> subscriber;
    Shelf shelf;
    String newValue;
    String cacheValue;

    @Before
    public void beforeEach() {
        shelf = new Shelf(new File("/tmp"));
        item = shelf.item("string");
        subscriber = new TestObserver<>();
        newValue = "new value";
        cacheValue = "cache value";
    }

    private void givenNoCache() {
        item.clear();
    }

    private void givenValidCache() {
        item.maxAge(5000).put(cacheValue);
    }

    private void givenInvalidCache() {
        item.maxAge(0, TimeUnit.MINUTES).put(cacheValue);
    }

    private void givenNoNew() {
        newValue = null;
    }

    private void printValues(List<String> values) {
        for (String value : values) System.out.println("value: " + value);
    }

    private void whenCacheThenNewSubscription() {
        Observable.just(newValue)
                .compose(item.cacheThenNew(String.class))
                .subscribe(subscriber);
    }

    @Test
    public void testNoCache() {
        givenNoCache();
        whenCacheThenNewSubscription();

        subscriber.assertValues(null, newValue);
        assertEquals(item.get(String.class), newValue);

        // TODO New TestObserver does not contain the getOnNextEvents method, nor a clear similar
        // method.  Call has been commented out to allow compiling, until such time as a replacement
        // is built into RxJava2.
        // printValues(subscriber.getOnNextEvents());
    }

    @Test
    public void testValidCache() {
        givenValidCache();
        whenCacheThenNewSubscription();

        subscriber.assertValues(cacheValue);
        assertEquals(item.get(String.class), cacheValue);
        // TODO New TestObserver does not contain the getOnNextEvents method, nor a clear similar
        // method.  Call has been commented out to allow compiling, until such time as a replacement
        // is built into RxJava2.
        // printValues(subscriber.getOnNextEvents());
    }

    @Test
    public void testInvalidCache() {
        givenInvalidCache();
        whenCacheThenNewSubscription();

        subscriber.assertValues(cacheValue, newValue);
        assertEquals(item.get(String.class), newValue);
        // TODO New TestObserver does not contain the getOnNextEvents method, nor a clear similar
        // method.  Call has been commented out to allow compiling, until such time as a replacement
        // is built into RxJava2.
        // printValues(subscriber.getOnNextEvents());
    }

    @Test public void testNoCacheAndNoNew() {
        givenNoCache();
        givenNoNew();
        whenCacheThenNewSubscription();

        subscriber.assertValue((String)null);
        assertEquals(item.get(String.class), null);
        // TODO New TestObserver does not contain the getOnNextEvents method, nor a clear similar
        // method.  Call has been commented out to allow compiling, until such time as a replacement
        // is built into RxJava2.
        // printValues(subscriber.getOnNextEvents());
    }

    @Test public void  testNoNewAndInvalidCache() {
        givenNoNew();
        givenInvalidCache();
        whenCacheThenNewSubscription();

        subscriber.assertValues(cacheValue);
        assertEquals(item.get(String.class), cacheValue);
        // TODO New TestObserver does not contain the getOnNextEvents method, nor a clear similar
        // method.  Call has been commented out to allow compiling, until such time as a replacement
        // is built into RxJava2.
        // printValues(subscriber.getOnNextEvents());
    }
}

File: shelf/src/test/java/com/toddway/shelf/CacheThenNewTests.kt
package com.toddway.shelf

import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import rx.Observable
import rx.observers.TestSubscriber
import java.io.File
import java.util.concurrent.TimeUnit

/**
 * Created by tway on 1/9/17.
 */

class CacheThenNewTests {

    internal lateinit var item: ShelfItem
    internal lateinit var subscriber: TestSubscriber<String>
    internal lateinit var shelf: Shelf
    internal var newValue: String? = null
    internal lateinit var cacheValue: String

    @Before
    fun beforeEach() {
        shelf = Shelf(File("/tmp"))
        item = shelf.item("string")
        subscriber = TestSubscriber()
        newValue = "new value"
        cacheValue = "cache value"
    }

    private fun givenNoCache() {
        item.clear()
    }

    private fun givenValidCache() {
        item.maxAge(5000).put(cacheValue)
    }

    private fun givenInvalidCache() {
        item.maxAge(0, TimeUnit.MINUTES).put(cacheValue)
    }

    private fun givenNoNew() {
        newValue = null
    }

    private fun printValues(values: List<String>) {
        for (value in values) println("value: " + value)
    }

    private fun whenCacheThenNewSubscription() {
        Observable.just<String>(newValue)
                .compose(item.cacheThenNew(String::class.java))
                .subscribe(subscriber)
    }

    @Test
    fun testNoCache() {
        givenNoCache()
        whenCacheThenNewSubscription()

        subscriber.assertValues(null, newValue)
        assertEquals(item[String::class.java], newValue)

        printValues(subscriber.onNextEvents)
    }

    @Test
    fun testValidCache() {
        givenValidCache()
        whenCacheThenNewSubscription()

        subscriber.assertValues(cacheValue)
        assertEquals(item[String::class.java], cacheValue)
        printValues(subscriber.onNextEvents)
    }

    @Test
    fun testInvalidCache() {
        givenInvalidCache()
        whenCacheThenNewSubscription()

        subscriber.assertValues(cacheValue, newValue)
        assertEquals(item[String::class.java], newValue)
        printValues(subscriber.onNextEvents)
    }

    @Test
    fun testNoCacheAndNoNew() {
        givenNoCache()
        givenNoNew()
        whenCacheThenNewSubscription()

        subscriber.assertValue(null)
        assertEquals(item[String::class.java], null)
        printValues(subscriber.onNextEvents)
    }

    @Test
    fun testNoNewAndInvalidCache() {
        givenNoNew()
        givenInvalidCache()
        whenCacheThenNewSubscription()

        subscriber.assertValues(cacheValue)
        assertEquals(item[String::class.java], cacheValue)
        printValues(subscriber.onNextEvents)
    }
}
